src/main/java/com/emc/object/s3/S3SignerV4.java

```diff
@@ -0,0 +1,397 @@
+            CanonicalURI + '\n' +
+            CanonicalQueryString + '\n' +
+            CanonicalHeaders + '\n' +
+            SignedHeaders + '\n' +
+            UNSIGNED-PAYLOAD
+         */
+        StringBuilder canonicalRequest = new StringBuilder();
+        canonicalRequest.append(method).append("\n");
+        String resource = RestUtil.getEncodedPath(uri);
+        canonicalRequest.append(resource).append("\n");
+        canonicalRequest.append(getCanonicalizedQueryString(parameters));
+
+        SortedMap<String, String> canonicalizedHeaders = getCanonicalizedHeaders(headers, parameters);
+        StringBuilder signedHeaders = new StringBuilder();
+        for (String name : canonicalizedHeaders.keySet()) {
+            canonicalRequest.append(name).append(":").append(canonicalizedHeaders.get(name).trim());
+            canonicalRequest.append("\n");
+            if (signedHeaders.length() != 0)
+                signedHeaders.append(";");
+            signedHeaders.append(name);
+        }
+        canonicalRequest.append("\n");
+
+        signedHeaders.append("\n");
+        canonicalRequest.append(signedHeaders);
+
+        canonicalRequest.append(S3Constants.AMZ_UNSIGNED_PAYLOAD);
+        log.debug("CanonicalRequest: {}", canonicalRequest);
+        return canonicalRequest.toString();
+    }
+
+    protected String getCanonicalRequest(ClientRequest request, Map<String, String> parameters, Map<String, List<Object>> headers) {
+        /*
+        CanonicalRequest =
+            HTTPRequestMethod + '\n' +
+            CanonicalURI + '\n' +
+            CanonicalQueryString + '\n' +
+            CanonicalHeaders + '\n' +
+            SignedHeaders + '\n' +
+            HexEncode(Hash(RequestPayload))
+         */
+        StringBuilder canonicalRequest = new StringBuilder();
+        canonicalRequest.append(request.getMethod()).append("\n");
+        URI uri = request.getURI();
+        String resource = RestUtil.getEncodedPath(uri);
+        canonicalRequest.append(resource).append("\n");
+        canonicalRequest.append(getCanonicalizedQueryString(parameters));
+
+        SortedMap<String, String> canonicalizedHeaders = getCanonicalizedHeaders(headers, parameters);
+        StringBuilder signedHeaders = new StringBuilder();
+        for (String name : canonicalizedHeaders.keySet()) {
+            canonicalRequest.append(name).append(":");
+            if (canonicalizedHeaders.get(name) != null) {
+                canonicalRequest.append(canonicalizedHeaders.get(name).trim());
+            }
+            canonicalRequest.append("\n");
+            if (signedHeaders.length() != 0)
+                signedHeaders.append(";");
+            signedHeaders.append(name);
+        }
+        canonicalRequest.append("\n");
+
+        signedHeaders.append("\n");
+        canonicalRequest.append(signedHeaders);
+
+        String hashedPayload = "";
+        String payload = "";
+        byte[] hash = hash256(payload);
```

Comment:
this will always be the same hash (of an empty string) - why is this necessary?

also, can we combine the two `getCanonicalRequest()` methods and add a boolean param whether to include the hash?  30 lines of duplicate code...

Reply:
1. Made changes to initializeaa static final value for hashed payload. 
2. Combined the two getCanonicalRequest() methods.